[Feat] Team04 - 이슈 조회 및 수정, 댓글 생성 및 수정 API 구현

longrunBiin opened the pull request:
1. 작업 내용
    - 이슈 상세 조회 API
    - 이슈 수정 API
        - 레이블, 담당자는 PUT
        - 제목, 내용, 이미지, 상태, 마일스톤은 PATCH
    - 에러 처리를 위한 ControllerAdvice 작성
    - Api 응답 통일을 위한 ApiResponse 작성
    - 댓글 생성 API
    - 댓글 수정 API
    - 쿼리 로직 검사가 필요한 테스트 H2 기반으로 리팩토링
    
2. 이번 주 달라진 점! 자랑할 점! 
    - (브리) : 댓글 생성/수정 기능을 추가하면서 파일 업로드 및 저장 로직에 대해 더 자세하게 이해할 수 있었습니다. 이전에 작성된 코드와 형식과 구조를 맞추기 위해 노력하면서  다른 사람의 코드를 조금 더 쉽게 읽을 수 있게 된 것 같습니다.
    
    - (빙빙) : 이슈를 수정 API를 만들때  매핑 테이블로 관리되는 레이블과 담당자는 해당 이슈에 관련된 모든 레코드를 지운 후 새로 insert를 시키는 방식을 써서 Put으로 정했습니다.
        
        반대로 매핑 테이블을 가지지 않는 값들은 Patch로 각각의 값만 수정하게끔 했습니다.
        
        기존의 ResponseEntity로 Api를 응답을 하면서 에러 상황에서의 응답이 유연하지 못한 불편함이 있었습니다. ApiResponse 클래스를 ResponseEntity에 담아 응답하게 해서 api 요청이 성공한 경우와 실패한 경우, 어떤 경우든 유연하게 필요한 데이터를 넘길 수 있도록 하였습니다.
        

1. 앞으로 진행할 작업
    - 버그 원인 찾기
    - 예약어와 겹치는 테이블명 고려해서 리팩토링
    - CI/CD를 위한 깃허브 액션 워크플로우 작성
    - 구현된 기능들 테스트 케이스 추가

1. 기타
    - 어제 라이브 코딩으로 말씀하신 리팩토링과 버그 수정 작업은 기능 구현하느라 아직 시작하지 못해서 금요일까지 마무리 해볼 예정입니다.
    - 댓글 수정 로직이 테스트(H2)에서는 성공하는데 PostMan으로는 오류가 발생(예약어 충돌)해서 금요일까지 리팩토링하도록 하겠습니다.

longrunBiin committed:
test : 존재하지 안흔ㄴ 이슈 조회 테스트
- 존재하지 안흔ㄴ 이슈를 조회하면 에러발생

longrunBiin committed:
feat : 이슈조회 컨트롤러 작성
- Api 을답을 위한 ApiResponse 작성
- ResponseDto 작성
- IssueNotFoundException 커스텀 예외 작성

longrunBiin committed:
test : IssueService.getIssueDetailById 테스트 작성
- 정상조회 테스트
- 존재하지 않는 아이디 테스트

longrunBiin committed:
chore : CommentResponseDto.CommentInfo 오타 수정
- commentInfo -> CommentInfo
- 에러 메세지 수정

longrunBiin committed:
feat : 이슈 상세조회 구현
- IssueService.getIssueDetailById 구현
- IssueDao.findIssueDetailById 작성

longrunBiin committed:
fix : sql 오타 수정
- 테이블 컬럼명 오타 수정

longrunBiin committed:
refactor : Controller Advice 구현
- IssueController.searchIssueDetail 전역 에러 처리
- IssueNotFoundException 수정

longrunBiin committed:
test : 이슈 세부 조회 테스트

moongyuxx committed:
test: H2 기반 테스트 환경 구성

moongyuxx committed:
test: 기존 mock 테스트 → H2 기반 테스트로 전환

moongyuxx committed:
fix: 이슈 카운트 조회 시 null 반환에 대한 방어 로직 추가

moongyuxx committed:
refactor: 동시성 반영
* ID 중 실제로 존재하는 이슈만 처리하고, 존재하지 않는 ID는 응답 메시지로 알려줌

moongyuxx committed:
test: 테스트에 H2반영

moongyuxx committed:
test: 이슈 상세 정보 테스트

longrunBiin committed:
Merge pull request #40 from codesquad-masters2025-team04/refactor/test

[REFACTOR] 테스트에 H2 DB 적용 및 리뷰 반영

longrunBiin committed:
feat : 이슈 제목, 내용, 이미지, 상태 수정 api 구현

longrunBiin committed:
feat : ControllerAdvice 커스템 에러 등록
- MilestoneNotFoundException
- FileUploadException

longrunBiin committed:
test : 이슈 업데이트 테스트 작성

longrunBiin committed:
fix : 이슈 업데이트 오류 수정
- 마일스톤 아이디 null 체크 추가
- 이미지 없는 경우 추가

longrunBiin committed:
chore : H2 컬럼 소문자 조회를 위한 @Column 추가
- H2에서 CrudRepository를 사용하는 경우 대문자로 컬럼을 찾아서 소문자로 변경

longrunBiin committed:
chore : CreateIssueDto 이름변경
- ApiMessageDto로 이름 변경

longrunBiin committed:
test : 이슈 레이블 수정 테스트 작성
- 정상 로직
- 존재하지 않는 이슈인 경우
- 존재하지 않는 레이블인 경우
- 레이블 리스트가 비어있는 경우

longrunBiin committed:
feat : 이슈 레이블 수정 기능 구현
- 이슈와 레이블이 존재하는지 확인
- 기존 매핑을 삭제
- 새 레이블 매필 추가
- CreateIssueDto 레이블, 담당자 Set으로 변경

longrunBiin committed:
chore : gitignore .idea 추가

moongyuxx committed:
feat: 댓글 생성 및 응답 DTO 추가

moongyuxx committed:
feat: 댓글 생성 서비스 및 저장소 구현

moongyuxx committed:
feat: 댓글 생성 컨트롤러 구현

moongyuxx committed:
test: 댓글 생성 서비스 단위 테스트 코드 추가

longrunBiin committed:
test : 이슈 담당자 수정 테스트 작성

longrunBiin committed:
feat : 이슈 담당자 수정 API 구현

longrunBiin committed:
refactor : deleteAllIssueLabelByIssueId() 클래스 이동
- IssueDao -> LabelDao로 이동

longrunBiin committed:
refactor : 커스텀 예외 getter제거
- getter 제거 후 message 사용
- 에러 메세지 상수 처리
- ControllerAdvice 간소화

longrunBiin committed:
refactor : 컨트롤레 예외처리 전역으로 처리

longrunBiin committed:
refactor : 컨트롤러 반환 타입 ResponseEntity<ApiResponse<?>>로 수정
- 로직 실패 시의 응답을 위해 반환 타입 수정

longrunBiin committed:
test : 컨트롤러 반환타입 변경에 따른 테스트 수정

moongyuxx committed:
feat: 댓글 수정 Dto 추가

moongyuxx committed:
feat: 댓글 수정 서비스 구현

moongyuxx committed:
feat: CommentNotFoundException 추가

moongyuxx committed:
feat:  댓글 수정 API 구현

moongyuxx committed:
test: 댓글 수정 테스트 추가

moongyuxx committed:
Merge pull request #41 from codesquad-masters2025-team04/feat/issues

[Feat] 이슈 수정 API 구현

moongyuxx committed:
feat: NOT_FOUND_COMMENT 상수 추가

longrunBiin committed:
Merge pull request #42 from codesquad-masters2025-team04/feat/comment

[FEAT] 댓글 기능 구현

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/comment/CommentRepository.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface CommentRepository extends CrudRepository<Comment, Long> {
+
+}
```
(개선)

저번 부터 25층이 계속 나오네요.
이 설정으로 자동으로 없앨 수 있습니다. (https://jojoldu.tistory.com/673)

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/comment/dto/CommentResponseDto.java:
```diff
@@ -0,0 +1,38 @@
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import software.amazon.awssdk.annotations.NotNull;
```
(짜잘)

사용하지 않는 import는 정리해주세요.
자동화 설정: https://backendcode.tistory.com/234
맥북 단축키: control + option + O
윈도우 단축키: ctrl + alt + O

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/entity/Comment.java:
```diff
@@ -13,11 +14,32 @@
 @Builder
 public class Comment {
     @Id
+    @Column("comment_id")
```
(질문)

칼럼 이름을 일일히 다 지정하신 이유는 뭔가요?

longrunBiin replied:
테스트 DB로 H2를 사용하게 되면서 테스트에서 사용된 쿼리 메소드로 만든 메소드들이 전부 컬럼을 대문자로 검색을 하게 되었습니다.
여러 설정들을 만져봤지만 해결되지 않아서 컬럼명을 소문자로 명시하는 방식을 사용했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/entity/Label.java:
```diff
@@ -13,10 +14,21 @@
+    @Column("updated_at")
     private LocalDateTime updatedAt;
-}
+}
```
(짜잘)

요기도 25F

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({IssueNotFoundException.class, MilestoneNotFoundException.class})
+    public ResponseEntity<ApiResponse<?>> handleNotFoundException(RuntimeException ex) {
```
(질문)

이 부분은 좀 더 나아질 수 있을거 같아요.
- ResponseEntity로 감싸는 이유는 뭔가요?
- 와일드 카드 대신 다른 걸 쓸 수는 없을까요?

longrunBiin replied:
ResponseEntity를 사용한 이유는 헤더에 상태 코드를 지정할 수 있고 바디를 일관된 구조로 응답 할 수 있을것 같아 사용했습니다.

와일드카드 대신 에러 응답 전용 DTO를 만들어서 사용할 수 있을 것 같습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueController.java:
```diff
@@ -28,51 +31,71 @@ public class IssueController {
 
     @GetMapping("")
-    public ResponseEntity<IssueResponseDto.IssueListDto> showIssueList(@RequestParam(name = "is_open") boolean isOpen, Pageable pageable) {
+    public ResponseEntity<ApiResponse<IssueResponseDto.IssueListDto>> showIssueList(@RequestParam(name = "is_open") boolean isOpen, Pageable pageable) {
```
(질문)

`GlobalExceptionHandler`에서 올린 질문과 비슷합니다.
- ResponseEntity로 꼭 감싸야할까요? 왜 감싸셨나요?

longrunBiin replied:
ResponseEntity를 안쓰고 ApiResponse에 상태 코드 필드를 추가하면 바디에만 코드가 적용되고 Http 상태 코드는 200 OK만 나가게 돼서 ResponseEntity로 감싸는 방식을 사용했습니다.

nag0124 replied:
그렇다면 예외를 RestControllerAdvice에서 처리하여 여러 http status를 보내도록 하고, 컨트롤러에서 보내는 답은 200OK로 내보내는건 어떤가요?

longrunBiin replied:
생각해보니 예외 상황 말고는 200OK로 충분한것 같습니다
컨트롤러 부분은 ApiResponse만 응답하도록 수정해놓겠습니다

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueDao.java:
```diff
@@ -31,12 +35,12 @@ public List<Map<String, Object>> findIssuesByOpenStatus(boolean isOpen, int page
                        a.profile_image AS assignee_profile
                 FROM issue i
-                LEFT JOIN user u ON i.author_id = u.user_id
+                LEFT JOIN `user` u ON i.author_id = u.user_id
```
(질문)

`user`가 예약어라 백틱으로 감싸신건가요?
mysql에서 테이블이 생성되지 않나요?

longrunBiin replied:
mysql에서는 잘 생성이 됐는데 테스트에서 h2를 사용하면서 테이블을 찾지 못하는 문제가 발생해 백틱을 사용했습니다.

moongyuxx replied:
mysql에서는 문제가 없지만, 테스트에서 사용하는 H2에서 `user`가 예약어라 오류가 발생했습니다.
일단 백틱으로 감싸서 해결습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueDao.java:
```diff
@@ -52,16 +56,43 @@ public int countIssuesByOpenStatus(boolean isOpen) {
+                    u.profile_image AS author_profile
+                FROM issue i
+                JOIN comment c ON i.issue_id = c.issue_id
+                JOIN user u ON u.user_id = c.author_id
```
(개선)

여기서도 `user`를 백틱으로 감싸서 통일성을 맞추는게 좋을거 같습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -135,33 +152,14 @@ public IssueResponseDto.CreateIssueDto createIssue(IssueRequestDto.CreateIssueDt
+            addNewLabels(issueId, request.getLabelId());
         }
 
         if(request.getAssigneeId() != null) {
```
(짜잘)

`request.getAssigneeId()`라고 지으니 담당자가 한명인 것처럼 느껴집니다. 변수에 s를 붙여도 좋을거 같아요.
위 부분의 `labelId`도 마찬가지입니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -135,33 +152,14 @@ public IssueResponseDto.CreateIssueDto createIssue(IssueRequestDto.CreateIssueDt
-                .id(issueId)
-                .message(CREATE_ISSUE)
-                .build();
+        return createMessageResult(issueId, CREATE_ISSUE);
```
(꼬리)

여기서 메세지를 보내주고 있는데, 의도가 무엇인가요?

longrunBiin replied:
이슈 생성을 성공한 경우에 명시적으로 알려주고 싶어서 메세지를 작성했습니다.

nag0124 replied:
지금 이슈트래커에서는 이슈를 만드는 데 성공하면 200 OK 응답을 내려주고 있네요.
'이슈를 만드는데 성공했습니다'라는 의미만 전달하려면 200 OK http status만 내려줘도 충분할 것 같습니다.
물론 생성된 값을 가지고 프론트에서 사용한다면 값이 전달되어야 하지만, 단지 성공을 알리는 거라면 200 OK만으로도 충분하지 않을까 싶어요.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
             throw new IssueStatusUpdateException(ExceptionMessage.NO_ISSUE_IDS);
```
(조언)

리스트가 없거나 비어있으면 **예외**를 던지고 있네요. 그런데 한번 생각해보셨으면 하는게 있어요.
- 유저가 빈 배열을 보내거나 null 값이면 이 것이 예외를 던질만한 사항인가?
- 처리할 것이 없으니 단순한 값으로 return해도 괜찮지 않나?
- 빈 값을 처리할 거면 Controller에서 값을 받을 때 값 검증을 하는건 어떤가?
- 궁극적으로 언제 반환값이 있고, 언제 단순한 return(처리할 것이 없어서 아무 값도 없는 응답)을 하고, 언제 예외를 던질 것인가?

좀 어려운 부분이긴한데 코드짜면서 응답을 어떻게 할지 고민하면 좋을것 같아서 올려봅니다.

moongyuxx replied:
* 이슈를 선택하지 않으면 프론트에서 상태 수정 요청을 보낼 수 없게 되어 있어서,
이슈 ID가 비어 있는 요청이 왔다면 프론트 버그거나 예외적인 상황이라고 판단해 에러로 처리했습니다.

*  Controller가 요청 데이터의 유효성을 책임지고, Service는 비즈니스 로직만 다루도록 역할을 분리하는 게 구조적으로 더 명확하기 때문에, 빈 요청을 서비스에서 검증하기보다는 컨트롤러 단에서 @Valid를 사용해 검증하는 방식이 더 좋은 것 같습니다.

* 정상 요청 / 처리할 게 없는 요청 / 잘못된 요청의 기준을 더 고민해보겠습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
         return IssueResponseDto.BulkUpdateIssueStatusDto.builder()
-                .issuesId(issueIds)
-                .message(UPDATE_ISSUESTATUS)
+                .issuesId(existingIds)
```
(질문)

존재하는 id만 필터링 해서 내려주고 있는데, 이 아이디를 프론트에서 사용하나요?

moongyuxx replied:
현재 프론트에서 사용하고 있지는 않습니다.
테스트에서 어떤 이슈가 상태 변경되었는지 검사하기 위해 포함시켰지만, 크게 필요한 것 같지는 않습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
 
-        Integer count = issueDao.countExistingIssuesByIds(issueIds, placeholders);
+        List<Long> missingIds = issueIds.stream()
+                .filter(id -> !existingIds.contains(id))
```
(개선)

`contains()`를 사용하실거면 `existingIds`를 가져올 떄 Set으로 가져오는게 나을거 같아요.
왜 그런지 List와 Set의 `contains()` 동작 방식을 비교해서 설명해주세요.

moongyuxx replied:
* List는 내부적으로 equals()를 맨 앞부터 순차적으로 비교하기 때문에 contains() 호출 시 **O(n)**의 시간이 걸립니다.

* Set은 해시 기반 탐색을 사용하여, hashCode()로 버킷 위치를 빠르게 찾고, 해당 버킷에서만 equals()로 비교하기 때문에 평균 O(1), 최악의 경우에도 O(log n) 시간 복잡도를 가집니다.

따라서 contains()를 사용할 경우, Set이 List보다 성능상 더 유리합니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
         return IssueCountDto.builder()
-                .openCount(openCount)
-                .closedCount(closedCount)
+                .openCount(openCount != null ? openCount : 0)
```
(개선)

여기서는 null이 들어갈 일이 없을 거 같군요?
왜 그럴까 생각하고 알려주세요.

moongyuxx replied:
현재 COUNT(*) 쿼리를 사용하고 있고, 이 경우는 결과가 없더라도 0이 반환되기 때문에 null이 들어갈 일이 없습니다.
일단 최대한 방어적으로 처리하기 위해 만들었는데, 반드시 필요한 로직은 아닌 것 같습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+    public searchIssueDetailDto getIssueDetailById(Long issueId) {
+        List<Map<String, Object>> issueById = issueDao.findIssueDetailById(issueId);
+        if (issueById.isEmpty()) {
+            throw new IssueNotFoundException(issueId);
```
(조언)

위에서 예외를 던지는 것에 관해서 조언을 해드렸는데, 이 경우와 위의 경우를 비교해보면 좋을거 같아요.

longrunBiin replied:
이슈를 수정하는 요청을 보냈는데 해당하는 이슈가 없다는 것은 이슈가 삭제된 후에 수정 요청이 들어왔거나 아예 존재하지 않는 이슈에 대한 수정 요청이라고 생각합니다. 
그런데 이 두가지 경우 전부 정상 로직 흐름은 아니라고 생각이 들어서 에러를 사용했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+                        .imageUrl((String) row.get("comment_image_url"))
+                        .createdAt(((Timestamp) row.get("comment_created_at")).toLocalDateTime())
+                        .author(UserDto.UserInfo.builder()
+                                .id((Long) row.get("author_id"))
```
(꼬리)

여기서 유저 아이디를 주는 이유가 무엇인가요? 프론트에서 아이디를 사용하여 요청을 보내는 로직이 있나요?

longrunBiin replied:
이전에 만들어 두었던 DTO를 재사용해서 아이디를 포함하고 있는데 닉네임이 유니크로 설정되어있기 때문에 아이디가 반드시 필요하지는 않을 것 같습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+                .content(issueContent)
+                .contentImageUrl(issueImage)
+                .comments(comments)
+                .commentSize(comments.size())
```
(질문)

여기서 코멘트의 사이즈를 내려주는 이유는 뭔가요? comments로 충분하지 않을까요?

longrunBiin replied:
프론트에서 comments.size()를 호출할지 백에서 호출할지 얘기를 했었는데 큰 이유는 없습니다.
단순히 백에서 크게 부담되는 작업이 아니라서 코멘트 사이즈를 응답해주고 프론트에서 값만 사용하는 방식으로 진행을 했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+        }
+
+        //기존 이미지를 삭제하는 것인지 확인
+        Optional<String> newImageUrl = determineNewImageUrl(request, uploadUrl, oldIssue);
+
+        Issue updated = oldIssue.toBuilder()
+                .title(request.getTitle() != null ? request.getTitle() : oldIssue.getTitle())
+                .content(request.getContent() != null ? request.getContent() : oldIssue.getContent())
+                .imageUrl(newImageUrl.orElse(null))
```
(개선)

여기서는 `Optional`을 반환하고 값이 없으면 `null`값을 넣고 있네요.
이럴가면 `Optional`을 안써도 괜찮지 않을까요...?

longrunBiin replied:
이미지가 없는 경우를 null이 아닌 ""로 통일시켜서 불필요한 Optional을 제거하고 null-safe하게끔 해보겠습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+                .orElseThrow(() -> new IssueNotFoundException(issueId));
+
+        //존재하는 담당자인지 확인
+        validateAssigneeIdsExist(assigneeIds);
```
(칭찬)

이렇게 계속 validation하는 거 좋네요! 이러면 미리 예외 상황을 방지할수 있어서 좋습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+        return createMessageResult(issueId, UPDATE_ISSUE_LABEL);
+    }
+
+    private static ApiMessageDto createMessageResult(Long id, String message) {
```
(개선)

이 함수의 접근자는 static이 아니여도 괜찮지 않을까요?

longrunBiin replied:
다른 서비스에서도 해당 메소드를 사용하는 경우가 많을것 같아 static을 사용했는데 현재는 궅이 드럴 필요가 없어보입니다.
만약에 다른 클래스에서 많이 사용하게 된다면 다른 유틸 클래스로 분리시켜서 사용하는게 더 좋아보입니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/user/UserDao.java:
```diff
@@ -23,4 +27,17 @@ public List<UserDto.UserInfo> findUserForFiltering() {
+        namedParameterJdbcTemplate.update(sql, params);
+    }
+
+    public List<Long> findExistingAssigneeIds(Set<Long> assigneeIds) {
```
(조언)

여기서 assignee는 유저와 이슈 사이의 관계가 맺어지면서 생성되는 말인데, UserDao에서 assignee를 쓰는게 부자연스러운 것 같아요.

longrunBiin replied:
만들면서 좀 찜찜하긴 했는데 각 도메인에 맞도록 분리하겠습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/response/ApiResponse.java:
```diff
@@ -0,0 +1,46 @@
+                .message(message)
+                .build();
+    }
+}
```
(짜잘)

EOF!

nag0124 commented on backend/src/main/resources/schema.sql:
```diff
@@ -51,6 +51,7 @@ CREATE TABLE comment (
                          content TEXT NOT NULL,
                          author_id BIGINT NOT NULL,
                          issue_id BIGINT NOT NULL,
+                         image_url VARCHAR(255),
```
(질문)

이건 그냥 진짜 궁금한건데, 제가 피그마 봤을 때는 파일 업로드로 봤거든요. 이게 이미지로 바뀐건가요?

longrunBiin replied:
파일이 맞는데 작성할 당시에 댓글에 이미지 추가되는걸 생각하다보니 그런 것 같습니다.
파일로 수정해놓겠습니다

moongyuxx replied:
파일 업로드 방식입니다. 대부분 이미지 파일을 업로드할 거라고 생각해서 image_url이라는 이름으로 정했는데, 혼란을 줄 수 있는 부분인 것 같습니다.

nag0124 commented on backend/src/test/java/codesquad/team4/issuetracker/util/TestDataHelper.java:
```diff
@@ -0,0 +1,40 @@
+        VALUES (?, ?, ?)
+    """, id, title, color);
+    }
+}
```
(짜잘)

EOF!

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/comment/CommentController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @PostMapping(value = "", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public ResponseEntity<ApiResponse<CommentResponseDto.CreateCommentDto>> createComment(
+            @RequestPart("comment") @Valid CommentRequestDto.CreateCommentDto request,
```
(칭찬)

오 validation 적용하셨군요. 굳입니다.
그런데 multipart에서도 `@valid` 정상 작동 하나요?

moongyuxx replied:
현재 import가 잘못되어 있었는데, jakarta.validation.constraints.NotNull을 사용하면 multipart 에서도 validation이 정상적으로 작동합니다.
Postman으로 null 값을 포함한 요청을 보냈을 때 400 Bad Request가 발생하는 것을 확인했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+                .body(ApiResponse.fail(ex.getMessage()));
+    }
+
+    @ExceptionHandler({FileUploadException.class, IssueStatusUpdateException.class})
+    public ResponseEntity<ApiResponse<?>> handleBadRequestException(RuntimeException ex) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(ApiResponse.fail(ex.getMessage()));
+    }
+    @ExceptionHandler(CommentNotFoundException.class)
+    public ResponseEntity<ApiResponse<?>> handleCommentNotFoundException(CommentNotFoundException ex) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND)
+                .body(ApiResponse.fail(ex.getMessage()));
```
(조언)

이렇게 예외를 관리하다보면 귀찮은 작업이 많아질 수 있어요.
이미 코드가 중복인거 같긴한데, 예외의 종류만 다르고 처리하는 방법, 메세지는 똑같네요.
공통화해서 처리하면 좋을거 같습니다!
왜냐면... 제가 가장 귀찮아 하는게 커스텀 예외 만드는 거거든요... 처음데 `RuntimeException`으로 처리하고 나중에 바꾸는데 예외 만드는 것 만큼 귀찮은게 없습니다.

moongyuxx replied:
네, 예외 핸들러는 중복이라 공통화하는 게 더 깔끔할 것 같습니다.

nag0124 commented on backend/src/test/resources/application-test.yml:
```diff
@@ -0,0 +1,10 @@
+spring:
+  datasource:
+    url: jdbc:h2:mem:testdb;DATABASE_TO_UPPER=false;MODE=MySQL
+    driver-class-name: org.h2.Driver
+    username: sa
+    password:
+  sql:
+    init:
+      mode: always
+      schema-locations: classpath:schema-test.sql
```
(칭찬)

테스트 프로필 좋네요.

nag0124 commented on backend/src/test/resources/schema-test.sql:
```diff
@@ -0,0 +1,81 @@
+                        email VARCHAR(255) NOT NULL UNIQUE,
+                        nickname VARCHAR(255) NOT NULL UNIQUE,
+                        profile_image VARCHAR(255),
+                        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+                        updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
```
(꼬리)

이건 H2에서 `ON UPDATE`를 지원 안해서 빼신 건가요?

moongyuxx replied:
네,  H2 모드를 MySQL로 설정해도 ON UPDATE CURRENT_TIMESTAMP는 지원하지 않아서 제거했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+    private Optional<String> determineNewImageUrl(IssueUpdateDto request, String uploadUrl, Issue oldIssue) {
+        String newImageUrl = oldIssue.getImageUrl();
+
+        if (Boolean.TRUE.equals(request.getRemoveImage())) {
```
(질문)

여기서 boolean이 Wrapper로 감싸져있는데요. Request를 받을 때 null값일 수 있나요?
기획상 null이 될 수 없다면 primitive 값으로 request의 필드를 변경해도 좋을거 같아요.
Wrapper 클래스면 null도 고려해야하니까요.

longrunBiin replied:
기획상으로는 제목 수정, 상태 수정, 내용 수정 버튼이 다  분리되어 있다보니 제목만 수정하는 경우엔 removeImage라는 값에 null이 들어올 수도 있을 것 같습니다

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/issue/IssueService.java:
```diff
@@ -173,38 +171,186 @@ public IssueResponseDto.BulkUpdateIssueStatusDto bulkUpdateIssueStatus(IssueRequ
+        if (foundIds.size() != labelIds.size()) {
+            Set<Long> missing = new HashSet<>(labelIds);
+            foundIds.forEach(missing::remove);
+            throw new LabelNotFoundException(missing);
```
(꼬리)

여기서 존재하지 않는 id들을 내려주는데, 프론트에서 이 아이디들로 어떤 처리를 하나요?

longrunBiin replied:
현재 프론트에서 따로 처리하는 작업은 없지만 어떤 아이다가 존재하지 않는지 알려 주는게 테스트를 하거나 사용자에게 알림을 주거나 한다면 도움이 될 것 같아서 추가했습니다.

nag0124 commented on backend/src/main/java/codesquad/team4/issuetracker/comment/CommentService.java:
```diff
@@ -0,0 +1,56 @@
+    @Transactional
+    public CommentResponseDto.UpdateCommentDto updateComment(Long commentId, CommentRequestDto.UpdateCommentDto request, String uploadUrl) {
+        //댓글 조회
+        Comment comment = commentRepository.findById(commentId)
+                .orElseThrow(() -> new CommentNotFoundException(commentId));
```
(칭찬)

검증하는 거 좋네요.
근데 JPA 냄세가 나는거 같기도 하고...

nag0124 reviewed (changes requested):
야직... 화요일이야...? 와 3일이나 더 출근해야하군요?? 아이 신나 ^^

그리고 제가 리뷰 쓸때 자신에게 칭찬하기를 부탁드렸는데 피드백을 안드렸군요 ㅎㅎ..;;
먼저 브리

> (브리) : 댓글 생성/수정 기능을 추가하면서 파일 업로드 및 저장 로직에 대해 더 자세하게 이해할 수 있었습니다. 이전에 작성된 코드와 형식과 구조를 맞추기 위해 노력하면서 다른 사람의 코드를 조금 더 쉽게 읽을 수 있게 된 것 같습니다

파일 업로드 기능을 그냥 쓰지 않고 이해하고 쓰셨군요! 좋습니다. 내가 쓰는 코드를 이해하고 쓰는건 매우 중요하요. 저도 잘 못하는 부분이긴 한데, 매우 좋습니다.
그리고 다른 사람(빙빙이겠죠 아마?)의 코드를 쉽게 읽는 건 매우 좋은 일입니다. 이러면 협업하는 속도가 빨라지거든요. 뇌를 덜쓰면서 서로 코드를 잘 이해할 수 있죠.

그리고 빙빙 
> 이이슈를 수정 API를 만들때 매핑 테이블로 관리되는 레이블과 담당자는 해당 이슈에 관련된 모든 레코드를 지운 후 새로 insert를 시키는 방식을 써서 Put으로 정했습니다.
반대로 매핑 테이블을 가지지 않는 값들은 Patch로 각각의 값만 수정하게끔 했습니다.

여기서 일괄 삭제후 일괄 생성하셨는데, 좋은 접근인 거 같아요. 다만 다 삭제 하고 다 만드는것과 일일히 필터링 해가면서 남길 것, 삭제할 것, 생성할 것을 분류하는 방법이 어떤 장단점이 있을지 생각해보는 것도 좋을거 같아요. 메서드 하나하나 신경을 쓰는거 같아서 보기 좋습니다.

>기존의 ResponseEntity로 Api를 응답을 하면서 에러 상황에서의 응답이 유연하지 못한 불편함이 있었습니다. ApiResponse 클래스를 ResponseEntity에 담아 응답하게 해서 api 요청이 성공한 경우와 실패한 경우, 어떤 경우든 유연하게 필요한 데이터를 넘길 수 있도록 하였습니다.

불편함을 느끼셔서 구조를 변경하셨군요! 매우 좋습니다.

그리고 몇가지 리뷰를 덧붙이면
### 1. 이미지 URL을 처리할 떄 ""과 null?
지금 로직에서 S3에 업로드를 실패하면 ""로 나오네요.
그런데 `IssueService.updateIssue()`를 호출하면 `imageUrl`이 null값이 들어갈 수 있지 않나요?
이렇게 되면 이미지가 없다는 걸 표현하는게 null, "" 두가지가 됩니다.
이 경우 빈 이미지를 처리할 떄 누구는 null 체크만 하고 누구는 `isBlank()`만 호출하게 된다면 에러가 터질 수 있어요. 이건 맞춰주세요.

### 2. `S3FileService`에서 업로드시 null 값?
`S3FileService`에서 파일을 업로드하면 `Optional`로 값을 리턴하는데, 파일 업로드 시 null값이 들어갈 수 있나요? 
null값이 들어갈 수 있으면 `Optional`로 감싸서 강제로 null 처리를 하는게 맞는데, null값이 들어갈 가능성이 없다면 그냥 String으로 반환해도 좋을 것 같습니다.
(그리고 파일 업로드하실 떄 Optional로 file 비었는지 체크하는 건 바꿔주시면 안될까요...? 그냥 눈에 밟혀서 그래요...)

longrunBiin committed:
chore : EOF 수정

longrunBiin committed:
refactor : imageUrl 없는 경우 빈문자로 통일
- S3FileService Optional null 체크 제거

longrunBiin committed:
refactor : CreteIssueDto 필드명 변경
- assigneeId -> assigneeIds
- labelId -> labelIds

longrunBiin committed:
chore : user 테이블 백틱 설정

longrunBiin committed:
chore : IssueService.createMessageResult static 제거

longrunBiin committed:
refactor : UserDao.deleteAllByIssueId 클래스 이동
- AssigneeDao로 메소드 이동

longrunBiin committed:
refactor : imageUrl -> fileUrl로 수정
- DDL image_url -> file_url 수정

longrunBiin committed:
refactor : 커스텀 예외 구조 변경
- DataNotFoundException, InvalidRequestException 추상 예외 클래스 작성
- ControllerAdvice에서 추상 예외들을 핸들링 하도록 수정
- 예외 패키지 구조 수정

longrunBiin committed:
refactor : GlobalExceptionHandler 반환타입 구체화

longrunBiin committed:
깃헙 액션 워크플로우 작성

longrunBiin committed:
Merge pull request #46 from codesquad-masters2025-team04/chore/cicd

[chore] 깃헙 액션 워크플로우 작성

longrunBiin committed:
워크플로우 수정

longrunBiin committed:
워크플로우 수정

longrunBiin committed:
Merge pull request #49 from codesquad-masters2025-team04/chore/cicd

[Chore] 워크플로우 수정

longrunBiin committed:
워크플로우 프론트 api url 추가

longrunBiin committed:
Merge pull request #51 from codesquad-masters2025-team04/chore/cicd

[Chore] 워크플로우 프론트 api url 추가

nag0124 commented:
여러분들이 요청에 대해서 성공했을 때 응답으로 메세지와 아이디를 넘겨 주는 경우가 많더라고요.
메세지와 아이디를 넘겨주는 이유를 물어봤는데, 응답이 `필요할 것 같아서`, `미래에 혹시 모르니까` 인 거 같아요.
근데 `필요할 것 같아서`의 필요성의 근거가 빈약하다고 느꼈어요.
그래서 이 필요성의 근거를 찾아보는 것도 좋을 것 같습니다.

longrunBiin committed:
워크플로우 수정

longrunBiin committed:
워크플로우 백엔드 수정

longrunBiin committed:
워크플로우 백, 프론트 분리

[Chore] 깃헙 액션 워크플로우 수정

longrunBiin committed:
fix : 레이블과 담당자가 없으면 출력 안되는 문제 수정
- IssueDao.findIssuesByOpenStatus Inner join -? left join으로 수정

longrunBiin committed:
refactor : IssueService.getIssues 리팩토링
- 코드 구조 개선 및 메소드 분리

longrunBiin committed:
fix : 이슈 리스트 조회 페이징 문제 해결
- IssueDao에서 limit offset 제거
- 결과 리스트에서 page, size로 페이지네이션

longrunBiin committed:
fix : 이슈 상세 조회 버그 수정
- 댓글이 없는 경우 글을 조회하지 못하던 문제 수정

longrunBiin committed:
refactor : 컨트롤러 반환값 수정
- ResponseEntity<ApiResponse<>> -> ApiResponse<>

longrunBiin committed:
리뷰내용 반영 리팩토링 및 버그 수정

[Fix] 리뷰내용 반영 리팩토링 및 버그 수정

nag0124 reviewed (approved):
고생하셨습니다~!
